src/draggable-list/index.tsx

```diff
@@ -0,0 +1,177 @@
+import { DNode } from '@dojo/framework/core/interfaces';
+import drag, { Position } from '@dojo/framework/core/middleware/drag';
+import { create, tsx } from '@dojo/framework/core/vdom';
+import { createICacheMiddleware } from '@dojo/framework/core/middleware/icache';
+import dimensions from '@dojo/framework/core/middleware/dimensions';
+import { ThemeProperties } from '@dojo/framework/core/middleware/theme';
+
+import theme from '../middleware/theme';
+import * as css from '../theme/default/draggable-list.m.css';
+import { uuid } from '@dojo/framework/core/util';
+import GlobalEvent from '../global-event';
+
+export function move<T = any>(items: T[], from: number, to: number): T[] {
+	const result = [...items];
+	const item = result[from];
+	result.splice(from, 1);
+	result.splice(to, 0, item);
+	return result;
+}
+
+export interface DraggableListProperties extends ThemeProperties {
+	/** Only allow items to be moved vertically on the y-axis */
+	lockAxis?: boolean;
+	/** Called when an item is moved to a new position */
+	onMove: (from: number, to: number) => void;
+}
+
+export interface DraggableListChild {
+	content: DNode;
+	key: string;
+}
+
+interface DraggableListState {
+	dragIndex?: number;
+	itemHeight?: number;
+	itemWidth?: number;
+	keyRoot?: string;
+	lastPosition?: Position;
+	startPosition?: Position;
+	visualIndex?: number;
+}
+
+const factory = create({
+	dimensions,
+	drag,
+	icache: createICacheMiddleware<DraggableListState>(),
+	theme
+})
+	.properties<DraggableListProperties>()
+	.children<DraggableListChild[]>();
+
+export const DraggableList = factory(function DraggableList({
+	children,
+	middleware: { dimensions, drag, icache, theme },
+	properties
+}) {
+	const themeCss = theme.classes(css);
+	const items = children();
+
+	const { lockAxis, onMove } = properties();
+	const keyRoot = icache.get('keyRoot');
+	const lastPosition = icache.get('lastPosition');
+	const visualIndex = icache.get('visualIndex');
+	const renderIndex = visualIndex && Math.min(items.length - 1, Math.max(0, visualIndex));
+
+	let dragIndex = icache.get('dragIndex');
+	let itemHeight = icache.get('itemHeight');
+	let itemWidth = icache.get('itemWidth');
+	let startPosition = icache.get('startPosition');
+
+	const buildKey = (key: string) => `${keyRoot}-${key}`;
```

Comment:
The `uuid` is intentionally used as a key root over `id` as a means of selectively re-rendering nodes that rely on this key root. `id` is not sufficient since there's no way to say "give me a new ID on the next render so I know all nodes that rely on this ID re-render."

Reply:
I guess i'm missing why you would need to selectively re-render nodes here then?